### PR TITLE
GOOWOO-655 Signal Detection for Sold 10 Items

### PR DESCRIPTION
**Pull request by zamanq**
### Changes proposed in this Pull Request:

Implements the Sold 10 Items detection.

**Sold10ItemsEvaluator**: Counts paid WooCommerce orders with revenue (total > 0) instead of completed orders only; uses HPOS and legacy post-meta queries with paid statuses from `wc_get_is_paid_statuses()`.

**Once per merchant**: Introduces SiteScopedNotificationEvaluatorInterface so sold-10-items trigger/dismiss state is stored site-wide in `gla_notifications_site_state`, not per admin user.

Closes https://linear.app/a8c/issue/GOOWOO-655/be-signal-detection-case-22-sold-10-items

### Screenshots:
N/A

### Detailed test instructions:

**Setup**: Ensure the store has fewer than 10 paid orders with revenue, or use a fresh test site.

- Create 10 WooCommerce orders with status processing or completed and a non-zero total.

- Go to WooCommerce → Marketing.

- Confirm the “Drive more sales with Google Ads” notification appears (congrats on first 10 sales).

- Refresh the page — notification should still show until dismissed.

- Dismiss the notification.

- Refresh — notification should not reappear.

- Log in as a second admin — confirm it also does not appear (site-wide dismiss).

**Negative case**: With fewer than 10 paid revenue orders, the notification should not appear.

### CODE REVIEW

File: src/Notification/NotificationService.php [hunk 1]
```diff
@@ -28,10 +29,15 @@ class NotificationService implements ContainerAwareInterface {
...
 	/**
-	 * The user_meta key under which the notifications state is stored.
+	 * The user_meta key under which per-user notifications state is stored.
 	 */
 	protected const STATE_META_KEY = 'gla_notifications_state';
 
+	/**
+	 * The site option key under which merchant-scoped notifications state is stored.
+	 */
+	protected const SITE_STATE_OPTION_KEY = 'gla_notifications_site_state';
```
Comment (asvinb) [inline src/Notification/NotificationService.php]: @zamanq `gla` prefixed options should be defined in `OptionsInterface` 👍

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php [hunk 2]
```diff
@@ -59,19 +62,89 @@ public function get_snooze_duration(): ?int {
...
+		return (int) $count >= $minimum;
+	}
+
+	/**
+	 * Returns WooCommerce order statuses that represent paid orders.
+	 *
+	 * @return string[]
+	 */
+	private function get_paid_order_statuses(): array {
+		if ( function_exists( 'wc_get_is_paid_statuses' ) ) {
```
Comment (asvinb) [inline src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @zamanq Can you add a note indicating as from which WC version which function is available please?

Comment (asvinb) [reply src/Notification/Evaluators/Sold10ItemsEvaluator.php]: Per convention, compatibility workarounds need a marker comment so the minimum WC version is searchable:
```php
// compatibility-code "WC < X.Y" -- wc_get_is_paid_statuses() added in X.Y
```

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php [hunk 3]
```diff
@@ -59,19 +62,89 @@ public function get_snooze_duration(): ?int {
...
+				return 'wc-' . $status;
+			},
+			$statuses
 		);
+		$status_placeholders = implode( ', ', array_fill( 0, count( $wc_statuses ), '%s' ) );
+		$query_limit         = $minimum + 1;
+
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.PreparedSQLPlaceholders.UnfinishedPrepare -- dynamic IN() placeholders.
+		if ( OrderUtil::custom_orders_table_usage_is_enabled() ) {
+			$count = $wpdb->get_var(
```
Comment (asvinb) [inline src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @zamanq Let's  just have the SQL query defined and then after the condition, we call `$wpdb->prepare` and `$wpdb->get_var` only once.

**Review by asvinb — CHANGES_REQUESTED** [review]

File: src/Notification/NotificationService.php (hunk 1 above)
Comment (zamanq) [reply src/Notification/NotificationService.php]: Defined it in OptionsInterface.

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php (hunk 2 above)
Comment (zamanq) [reply src/Notification/Evaluators/Sold10ItemsEvaluator.php]: Note added.

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php (hunk 3 above)
Comment (zamanq) [reply src/Notification/Evaluators/Sold10ItemsEvaluator.php]: Refactored the SQL. While at it, I also refactored the SQL for SalesNotGrowingEvaluator as well.

File: src/Notification/Evaluators/SalesNotGrowingEvaluator.php [hunk 4]
```diff
@@ -109,39 +109,35 @@ protected function get_gmv_for_period( DateTime $start, DateTime $end ): float {
...
-					WHERE type = 'shop_order'
-						AND status = %s
-						AND date_created_gmt >= %s
-						AND date_created_gmt <= %s",
-					'wc-completed',
-					$start_sql,
-					$end_sql
-				)
-			);
+			$query = "SELECT COALESCE( SUM( total_amount ), 0 )
```
Comment (asvinb) [inline src/Notification/Evaluators/SalesNotGrowingEvaluator.php:154]: @zamanq Curious, can we use `wc_get_orders` here instead of querying the table directly?
https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-order-functions.php#L37

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php [hunk 5]
```diff
@@ -59,19 +62,87 @@ public function get_snooze_duration(): ?int {
...
 
 	/**
-	 * Count completed orders up to the query limit.
+	 * Site-scoped evaluators share one cache entry per store.
 	 *
-	 * @return int
+	 * @return string
+	 */
+	protected function get_cache_key(): string {
+		return 'gla_notif_' . $this->get_id();
```
Comment (asvinb) [inline src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @zamanq I see we have a couple of places where we are using the `gla_notif_` prefix. Instead of hardcoding it across evaluators, can we look into defining it as a constant once and use it across all evaluators?

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php [hunk 6]
```diff
@@ -59,19 +62,87 @@ public function get_snooze_duration(): ?int {
...
+			static function ( string $status ): string {
+				return 'wc-' . $status;
+			},
+			$statuses
+		);
+		$status_placeholders = implode( ', ', array_fill( 0, count( $wc_statuses ), '%s' ) );
+		$query_limit         = $minimum + 1;
+
+		if ( OrderUtil::custom_orders_table_usage_is_enabled() ) {
+			$query = "SELECT COUNT(*) FROM (
```
Comment (asvinb) [inline src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @zamanq Same as my earlier comment, can we use `wc_get_orders` please?

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php (hunk 2 above)
Comment (asvinb) [reply src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @zamanq I think we can remove the guard in this case since it was introduced in WC 3.0, which is quite far behind. We can safely assume that `wc_get_is_paid_statuses` is available.

**Review by asvinb — CHANGES_REQUESTED** [review]

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php (hunk 3 above)
Comment (asvinb) [reply src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @zamanq See https://github.com/woocommerce/google-listings-and-ads/pull/3546#discussion_r3500202100

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php (hunk 2 above)
Comment (zamanq) [reply src/Notification/Evaluators/Sold10ItemsEvaluator.php]: Agreed. Removed the guard.

File: src/Notification/Evaluators/SalesNotGrowingEvaluator.php (hunk 4 above)
Comment (zamanq) [reply src/Notification/Evaluators/SalesNotGrowingEvaluator.php:154]: @asvinb `wc_get_orders()` is a good fit when you only need to find or count orders, like checking whether at least one completed order exists. It is not a good fit for this evaluator because we need the total GMV for a date range, and wc_get_orders() does not support aggregation.

To calculate GMV with it, we would have to fetch every completed order in the period and sum each order total in PHP. This evaluator does that twice per check: once for the current month and once for the same month last year. On stores with a meaningful order volume, that is much slower and uses far more memory than the current approach.

The direct SQL query does the same work in one aggregated SUM query per period. The tradeoff is that we need separate HPOS and legacy SQL, but that is worth it here for performance.

So the reasoning for keeping the direct query is: `wc_get_orders()` cannot sum order totals efficiently, and this feature needs exactly that. Let me know if you feel otherwise, we can then revisit this if needed :)

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php (hunk 5 above)
Comment (zamanq) [reply src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @asvinb Makes sense, done. Thank you :)

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php (hunk 6 above)
Comment (zamanq) [reply src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @asvinb Yes, for this case `wc_get_orders()` is reasonable here. Refactored.

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php [hunk 7]
```diff
@@ -59,19 +62,42 @@ public function get_snooze_duration(): ?int {
...
+	 * @return bool
 	 */
-	protected function get_completed_order_count(): int {
+	protected function has_minimum_revenue_orders( int $minimum ): bool {
+		$statuses = wc_get_is_paid_statuses();
+		if ( empty( $statuses ) ) {
+			return false;
+		}
+
 		$orders = wc_get_orders(
```
Comment (asvinb) [inline src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @zamanq 
The `'total' => ['value' => 0, 'operator' => '>']` arg works in HPOS but silently misbehaves in legacy post-meta storage.
* HPOS path: `OrdersTableQuery` maps `total` → `total_amount` and routes it through `generate_total_query()`, which has explicit support for the `['value' => ..., 'operator' => '>']` format. Emits `WHERE total_amount > 0`. ✅
* Legacy CPT path: `WC_Order_Data_Store_CPT::get_wp_query_args()` remaps `total` → `order_total`, then the parent `WC_Data_Store_WP` sees `_order_total` in `internal_meta_keys` and builds a generic meta query. Because the value is an array, it sets `compare => 'IN'`, producing effectively `_order_total IN (0, '>')`. That matches only zero-total orders, the opposite of what we want. ❌

I suggest explicit branching to match the pattern already in `SalesNotGrowingEvaluator`:
```php
$query_args = [
    'status' => $statuses,
    'limit'  => $minimum + 1,
    'return' => 'ids',
];

if ( OrderUtil::custom_orders_table_usage_is_enabled() ) {
    $query_args['total'] = [ 'value' => 0, 'operator' => '>' ];
} else {
    $query_args['meta_query'] = [
        [ 'key' => '_order_total', 'value' => 0, 'compare' => '>', 'type' => 'NUMERIC' ],
    ];
}

$orders = wc_get_orders( $query_args );
```

**Review by asvinb — CHANGES_REQUESTED** [review]

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php (hunk 7 above)
Comment (zamanq) [reply src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @asvinb Thanks! Adjusted accordingly

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php [hunk 8]
```diff
@@ -59,19 +63,57 @@ public function get_snooze_duration(): ?int {
...
+			return false;
+		}
+
+		$query_args = [
+			'status' => $statuses,
+			'limit'  => $minimum + 1,
+			'return' => 'ids',
+		];
+
+		// The 'total' => [ 'value' => 0, 'operator' => '>' ] shorthand is only honoured
```
Comment (asvinb) [inline src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @zamanq 
`OrdersTableQuery` doesn't support `'total' => ['value' => ..., 'operator' => ...]` for native columns — that shape is only meaningful for `meta_query`. The `operator` key is silently dropped and the comparison is hardcoded to `=`:
https://github.com/woocommerce/woocommerce/blob/4150acd12f4946f331b4c8ad357952f4fa18d0e0/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableQuery.php#L1112-L1137

Since the value is an array, `where()` upgrades `=` to `IN` and formats each element as `decimal` — `0` → `'0'`, and `'>'` → `''` → `0`:
https://github.com/woocommerce/woocommerce/blob/4150acd12f4946f331b4c8ad357952f4fa18d0e0/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableQuery.php#L1066-L1104

Resulting SQL is `total_amount IN (0, 0)` — i.e. `total_amount = 0`, the opposite of "revenue > 0." On any HPOS store (WC default since 8.2+), this filters for zero-total orders instead of revenue orders, so the notification won't fire correctly. The comment above (L96-98) also has it backwards — the legacy `meta_query` branch is the one that correctly filters `total > 0`.

**Fix:** no native numeric-range arg exists for order-table columns here. Either filter `wc_get_order( $id )->get_total() > 0` in PHP after fetching bounded IDs, or run direct SQL like `SalesNotGrowingEvaluator` does elsewhere in this PR.

Also: `has_minimum_revenue_orders()` is fully mocked in tests, so this query path has no coverage — worth an integration test with real zero/non-zero orders under both storage modes.

**Review by asvinb — CHANGES_REQUESTED** [review]

File: src/Notification/Evaluators/Sold10ItemsEvaluator.php (hunk 8 above)
Comment (zamanq) [reply src/Notification/Evaluators/Sold10ItemsEvaluator.php]: @asvinb Thank you for this! Makes sense. Refactored accordingly.

**Review by asvinb — APPROVED** [review]

**Comment by ankitguptaindia** [comment]
### QA Status: Approved ✅ 

#### Test Result: 



* Verified the **Drive More Sales with Google Ads** notification appears after the store reaches **10 paid orders with revenue**. ✅
* Confirmed the notification remains visible after a page refresh until it is dismissed. ✅
* Verified the notification does not reappear after being dismissed. ✅
* Verified the notification is not displayed when the store has fewer than **10 paid revenue orders**. ✅
* All tested scenarios passed successfully and behaved as expected. ✅


Screenshot 2026-07-16 at 6 02 24 PM
